hail/src/main/scala/is/hail/expr/ir/functions/CallFunctions.scala

```diff
@@ -1,38 +1,44 @@
   def registerAll() {
-    registerWrappedScalaFunction("Call", TString, TCall, null)(Call.getClass, "parse")
+    registerWrappedScalaFunction("Call", TString, TCall, (_: PType) => PCanonicalCall())(Call.getClass, "parse")
 
-    registerScalaFunction("Call", Array(TBoolean), TCall, null)(Call0.getClass, "apply")
+    registerScalaFunction("Call", Array(TBoolean), TCall, (pt: Seq[PType]) => PCanonicalCall())(Call0.getClass, "apply")
 
-    registerScalaFunction("Call", Array(TInt32, TBoolean), TCall, null)(Call1.getClass, "apply")
+    registerScalaFunction("Call", Array(TInt32, TBoolean), TCall, (pt: Seq[PType]) => PCanonicalCall())(Call1.getClass, "apply")
 
-    registerScalaFunction("Call", Array(TInt32, TInt32, TBoolean), TCall, null)(Call2.getClass, "apply")
+    registerScalaFunction("Call", Array(TInt32, TInt32, TBoolean), TCall, (pt: Seq[PType]) => PCanonicalCall())(Call2.getClass, "apply")
 
-    registerScalaFunction("UnphasedDiploidGtIndexCall", Array(TInt32), TCall, null)(Call2.getClass, "fromUnphasedDiploidGtIndex")
+    registerScalaFunction("UnphasedDiploidGtIndexCall", Array(TInt32), TCall, (pt: Seq[PType]) => PCanonicalCall())(Call2.getClass, "fromUnphasedDiploidGtIndex")
 
-    registerWrappedScalaFunction("Call", TArray(TInt32), TBoolean, TCall, null)(CallN.getClass, "apply")
+    registerWrappedScalaFunction("Call", TArray(TInt32), TBoolean, TCall, {
+      case(_: PType, _: PType) => PCanonicalCall()
+    })(CallN.getClass, "apply")
 
     val qualities = Array("isPhased", "isHomRef", "isHet",
       "isHomVar", "isNonRef", "isHetNonRef", "isHetRef")
-    for (q <- qualities) registerScalaFunction(q, Array(TCall), TBoolean, null)(Call.getClass, q)
+    for (q <- qualities) registerScalaFunction(q, Array(TCall), TBoolean, (pt: Seq[PType]) => PBoolean())(Call.getClass, q)
 
-    registerScalaFunction("ploidy", Array(TCall), TInt32, null)(Call.getClass, "ploidy")
+    registerScalaFunction("ploidy", Array(TCall), TInt32, (pt: Seq[PType]) => PBoolean())(Call.getClass, "ploidy")
 
-    registerScalaFunction("nNonRefAlleles", Array(TCall), TInt32, null)(Call.getClass, "nNonRefAlleles")
+    registerScalaFunction("nNonRefAlleles", Array(TCall), TInt32, (pt: Seq[PType]) => PInt32())(Call.getClass, "nNonRefAlleles")
 
-    registerScalaFunction("unphasedDiploidGtIndex", Array(TCall), TInt32, null)(Call.getClass, "unphasedDiploidGtIndex")
+    registerScalaFunction("unphasedDiploidGtIndex", Array(TCall), TInt32, (pt: Seq[PType]) => PInt32())(Call.getClass, "unphasedDiploidGtIndex")
 
-    registerScalaFunction("index", Array(TCall, TInt32), TInt32, null)(Call.getClass, "alleleByIndex")
+    registerScalaFunction("index", Array(TCall, TInt32), TInt32, (pt: Seq[PType]) => PInt32())(Call.getClass, "alleleByIndex")
 
-    registerScalaFunction("downcode", Array(TCall, TInt32), TCall, null)(Call.getClass, "downcode")
+    registerScalaFunction("downcode", Array(TCall, TInt32), TCall, (pt: Seq[PType]) => pt(0))(Call.getClass, "downcode")
```

Comment:
I think you will ask this to be PCanonicalCall. I have the same concerns as before, I think it is better to leave the constructible representation to the type, and not make those choices in each consumer of the type.